src/corefx/System.Globalization.Native/pal_collation.c

```diff
@@ -385,38 +378,40 @@ void GlobalizationNative_CloseSortHandle(SortHandle* pSortHandle)
         }
     }
 
-    pthread_mutex_destroy(&pSortHandle->collatorsLockObject);
-
     free(pSortHandle);
 }
 
 const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32_t options, UErrorCode* pErr)
 {
-    UCollator* pCollator;
     if (options == 0)
     {
-        pCollator = pSortHandle->collatorsPerOption[0];
+        return pSortHandle->collatorsPerOption[0];
     }
     else
     {
-        int lockResult = pthread_mutex_lock(&pSortHandle->collatorsLockObject);
-        if (lockResult != 0)
+        options &= CompareOptionsMask;
+        UCollator* pCollator = pSortHandle->collatorsPerOption[options];
+        if (pCollator != NULL)
         {
-            assert(FALSE && "Unexpected pthread_mutex_lock return value.");
+            return pCollator;
         }
 
-        options &= CompareOptionsMask;
-        pCollator = pSortHandle->collatorsPerOption[options];
-        if (pCollator == NULL)
+        pCollator = CloneCollatorWithOptions(pSortHandle->collatorsPerOption[0], options, pErr);
+        UCollator* pNull = NULL;
+
+#if HAVE_WORKING_STDATOMIC
+        if (!atomic_compare_exchange_strong((_Atomic(UCollator*)*)&pSortHandle->collatorsPerOption[options], &pNull, pCollator))
+#else
+        if (!__atomic_compare_exchange_n(&pSortHandle->collatorsPerOption[options], &pNull, pCollator, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
```

Comment:
I think it would be better to just use `__atomic_compare_exchange_n` all the time, just like corefx does. And add a comment that we are not using the `atomic_compare_exchange_strong` to workaround bugs in clang 5.0.

Reply:
The part I do not like is the code duplication - we are likely going to use this in more places over time and having this ifdef around every use would not be pretty. Another way to fix the code duplication is to:

 ```
#undef atomic_compare_exchange_strong  
#define atomic_compare_exchange_strong(obj, expected, val) __atomic_compare_exchange_n(obj, expected, val, false,  __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST)
```

to "fix" the broken `_Atomic` until we get a fixed compilers; and then use the standard `atomic_compare_exchange_strong` without anyifdefs.